interface/ceed-preconditioning.c

```diff
@@ -2465,12 +2465,35 @@ int CeedOperatorLinearAssemble(CeedOperator op, CeedVector values) {
 
     CeedCall(CeedOperatorGetNumElements(op, &num_elem));
     if (num_elem == 0) return CEED_ERROR_SUCCESS;
+    has_linear_assemble_single = op->LinearAssembleSingle != NULL;
+  } else {
+    CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
+    CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
+    has_linear_assemble_single = true;
+    for (CeedInt i = 0; i < num_suboperators; i++) {
+      has_linear_assemble_single = has_linear_assemble_single && sub_operators[i]->LinearAssembleSingle != NULL;
+    }
   }
 
   if (op->LinearAssemble) {
     // Backend version
     CeedCall(op->LinearAssemble(op, values));
     return CEED_ERROR_SUCCESS;
+  } else if (has_linear_assemble_single) {
+    // Default to summing contributions of suboperators
+    CeedCall(CeedVectorSetValue(values, 0.0));
+    if (is_composite && num_suboperators > 0 && sub_operators[0]) {
+      CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
+      CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
+      for (CeedInt k = 0; k < num_suboperators; k++) {
+        CeedCall(CeedSingleOperatorAssemble(sub_operators[k], offset, values));
+        CeedCall(CeedSingleOperatorAssemblyCountEntries(sub_operators[k], &single_entries));
+        offset += single_entries;
+      }
+    } else {
+      CeedCall(CeedSingleOperatorAssemble(op, offset, values));
+    }
+    return CEED_ERROR_SUCCESS;
   } else {
```

Comment:
```suggestion
  } else if (is_composite && has_linear_assemble_single) {
    // Default to summing contributions of suboperators
    CeedCall(CeedVectorSetValue(values, 0.0));
    if (num_suboperators > 0 && sub_operators[0]) {
      CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
      CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
      for (CeedInt k = 0; k < num_suboperators; k++) {
        CeedCall(CeedSingleOperatorAssemble(sub_operators[k], offset, values));
        CeedCall(CeedSingleOperatorAssemblyCountEntries(sub_operators[k], &single_entries));
        offset += single_entries;
      }
    }
    return CEED_ERROR_SUCCESS;
  } else {
```

Would this be better @jeremylt

Reply:
It's possible that none of the suboperators have `LinearAssembleSingle`, but the composite operator does have `LinearAssemble`, right?

Reply:
In that case we want to fallback and use the `LinearAssemble` before we try (and fail) to use single operator assemble

Reply:
Currently we only have implementations of `LinearAssembleSingle` and none of `LinearAssemble`.

We really, really want to avoid gen going down to ref, as that blows up our memory usage. We could find ourselves in a situation with the code in this PR where only some of the suboperators of the composite operator support `LinearAssembleSingle` in gen, and I'd like as many of those suboperators to use the native gen assembly as possible instead of falling back to ref preemptively

Reply:
Really, since assembly is much less frequent than operator application, I'm less worried about handling this as one big composite operation as we've got bigger perf targets that should be more impactful

Reply:
Alright, just updated. lmk what you think

Reply:
Close, I think we need this?
```C
int CeedOperatorLinearAssemble(CeedOperator op, CeedVector values) {
  bool          is_composite;
  CeedInt       num_suboperators, offset = 0;
  CeedSize      single_entries = 0;
  CeedOperator *sub_operators;

  CeedCall(CeedOperatorCheckReady(op));
  CeedCall(CeedOperatorIsComposite(op, &is_composite));

  // Early exit for empty operator
  if (!is_composite) {
    CeedInt num_elem = 0;
    CeedCall(CeedOperatorGetNumElements(op, &num_elem));
    if (num_elem == 0) return CEED_ERROR_SUCCESS;
  }

  if (op->LinearAssemble) {
    // Backend version
    CeedCall(op->LinearAssemble(op, values));
    return CEED_ERROR_SUCCESS;
  } else if (is_composite) {
    // Default to summing contributions of suboperators
    CeedCall(CeedVectorSetValue(values, 0.0));
    CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
    CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
    for (CeedInt k = 0; k < num_suboperators; k++) {
      CeedCall(CeedSingleOperatorAssemble(sub_operators[k], offset, values));
      CeedCall(CeedSingleOperatorAssemblyCountEntries(sub_operators[k], &single_entries));
      offset += single_entries;
    }
    return CEED_ERROR_SUCCESS;
  } else {
    // Operator fallback
    CeedOperator op_fallback;

    CeedCall(CeedOperatorGetFallback(op, &op_fallback));
    if (op_fallback) {
      CeedCall(CeedOperatorLinearAssemble(op_fallback, values));
      return CEED_ERROR_SUCCESS;
    }
  }

  // Default to interface version if non-composite and no fallback
  CeedCall(CeedVectorSetValue(values, 0.0));
  CeedCall(CeedSingleOperatorAssemble(op, offset, values));
  return CEED_ERROR_SUCCESS;
}
```

Reply:
the path is then:
- no `op->LinearAssemble`
- not composite
- there is an `op_fallback`, so call `CeedOperatorLinearAssemble(op_fallback, values)`
- `op_fallback` throws an error because it doesn't have a valid `LinearAssemble`

Reply:
Oh, duh
```C
int CeedOperatorLinearAssemble(CeedOperator op, CeedVector values) {
  bool          is_composite;
  CeedInt       num_suboperators, offset = 0;
  CeedSize      single_entries = 0;
  CeedOperator *sub_operators;

  CeedCall(CeedOperatorCheckReady(op));
  CeedCall(CeedOperatorIsComposite(op, &is_composite));

  // Early exit for empty operator
  if (!is_composite) {
    CeedInt num_elem = 0;

    CeedCall(CeedOperatorGetNumElements(op, &num_elem));
    if (num_elem == 0) return CEED_ERROR_SUCCESS;
  }

  if (op->LinearAssemble) {
    // Backend version
    CeedCall(op->LinearAssemble(op, values));
    return CEED_ERROR_SUCCESS;
  } else if (is_composite) {
    // Default to summing contributions of suboperators
    CeedCall(CeedVectorSetValue(values, 0.0));
    CeedCall(CeedCompositeOperatorGetNumSub(op, &num_suboperators));
    CeedCall(CeedCompositeOperatorGetSubList(op, &sub_operators));
    for (CeedInt k = 0; k < num_suboperators; k++) {
      CeedCall(CeedSingleOperatorAssemble(sub_operators[k], offset, values));
      CeedCall(CeedSingleOperatorAssemblyCountEntries(sub_operators[k], &single_entries));
      offset += single_entries;
    }
    return CEED_ERROR_SUCCESS;
  } else if (op->LinearAssembleSingle) {
    CeedCall(CeedVectorSetValue(values, 0.0));
    CeedCall(CeedSingleOperatorAssemble(op, offset, values));
    return CEED_ERROR_SUCCESS;
  } else {
    // Operator fallback
    CeedOperator op_fallback;

    CeedCall(CeedOperatorGetFallback(op, &op_fallback));
    if (op_fallback) {
      CeedCall(CeedOperatorLinearAssemble(op_fallback, values));
      return CEED_ERROR_SUCCESS;
    }
  }

  // Default to interface version if non-composite and no fallback
  CeedCall(CeedVectorSetValue(values, 0.0));
  CeedCall(CeedSingleOperatorAssemble(op, offset, values));
  return CEED_ERROR_SUCCESS;
}
```
there we go, this should be better

Reply:
yep that works, thanks!